runtime/bcverify/staticverify.c

```diff
@@ -1322,17 +1316,58 @@ checkStackMap (J9CfrClassFile* classfile, J9CfrMethod * method, J9CfrAttributeCo
 					}
 					if (frameType >= CFR_STACKMAP_CHOP_3) {
 						slotCount = (IDATA) frameType - CFR_STACKMAP_APPEND_BASE;
-						localSlots += slotCount;
 						if (slotCount < 0) {
 							slotCount = 0;
 						}
+
+						/* JCK 4.7.4: It is an error if, for any index i, locals[i] represents a local variable whose index is greater than the maximum 
+						 * number of local variables for the method. */
+						if ((frameType >= CFR_STACKMAP_APPEND_1) && (frameType <= CFR_STACKMAP_APPEND_3)) {
```

Comment:
Given that checkStackMapEntries is already used to validate the entries in stackmaps, it is better to leverage this function for `CFR_STACKMAP_APPEND`.

The code could be something like:
```
//add a new argument that validates the slots against the maxLocal or maxStack
checkStackMapEntries(classfile, code, map, &entries, slotCount, end, maxValue) {
int actualSlots = 0;
for (; slotCount; slotCount--) {
...
// only long/double data takes 2 slots
if ((entryType == CFR_STACKMAP_TYPE_LONG)
|| ((entryType == CFR_STACKMAP_TYPE_DOUBLE)
) {
  actualSlots += 2;
} else {
   actualSlots += 1;
 }
} //for

// check the accumulated values against maxLocals or maxStacks
if (actualSlots > maxValue) {
    return FATAL_CLASS_FORMAT_ERROR;
}
}

checkStackMap()
{
if ((frameType >= CFR_STACKMAP_APPEND_1) && (frameType <= CFR_STACKMAP_APPEND_3)) {
   ...
   errorCode = checkStackMapEntries (classfile, code, map, &entries, slotCount, end, code->maxLocals);
}
...
} else {
					/* full frame */
...
					errorCode = checkStackMapEntries (classfile, code, map, &entries, slotCount, end, code->maxLocals); <-----
 ...
				errorCode = checkStackMapEntries (classfile, code, map, &entries, slotCount, end, code->maxStack); <-------
...
}
```

Reply:
@ChengJin01 I've refactored the change.

Reply:
The VM Spec at 4.10.1.3 Instruction Representation says:
```
The length of the operand stack must not exceed the declared maximum size.
operandStackHasLegalLength(Environment, OperandStack) :
    length(OperandStack, Length),
    maxOperandStackLength(Environment, MaxStack),
    Length =< MaxStack.
```
So you still need to check the `slotCount` against `code->maxStack` (this is not for `CFR_STACKMAP_APPEND `but the existing code)  as `checkStackMapEntries `is used to check both `locals` and `stack`.

Even though `code `is passed as argument to checkStackMapEntries, the argument `maxValue `can be used for both  `code->maxLocals` and `code->maxStack`, in which case there is no need to introduce anther argument or flag to determine whether the passed-in argument is intended for `locals `or `stack` as mentioned in the pseudo-code above.
```
checkStackMapEntries(classfile, code, map, &entries, slotCount, end, maxValue) {
...
// check the accumulated values against maxLocals or maxStacks
if (actualSlots > maxValue) {
    return FATAL_CLASS_FORMAT_ERROR;
}
...
}
```

Reply:
> So you still need to check the slotCount against code->maxStack

It looks like this is already done for full-frame outside of checkStackMapEntries https://github.com/eclipse/openj9/blob/19571adf476a7080fa4cbbab0c0de3cb3e96f190/runtime/bcverify/staticverify.c#L1313

> there is no need to introduce anther argument or flag to determine whether the passed-in argument

passing maxLocals/maxStack will work fine as well. I don't see that solution is necessarily simpler as I would need to add another check to make sure that checkStackMapEntries is not called twice for the `CFR_STACKMAP_APPEND`.

Reply:
`if (code->maxStack < slotCount) `{ 
<---- this is only checked when` frameType != CFR_STACKMAP_FULL`
```
if (frameType != CFR_STACKMAP_FULL) {
...
/* The stackmap entry is invalid if the size of stack (1 slot) exceeds the size of the max stack.*/
	if (code->maxStack < slotCount) {
```

Reply:
@ChengJin01 updated